### DISCUSSION TITLE
If I use Skia4Delphi in a Windows form DLL project, do I need to include any files?

### DISCUSSION THREAD

**Discussion by War3Evo** [skia4delphi/skia4delphi#176] [Q&A]
If I use Skia4Delphi in a Windows form DLL project, do I need to include any files?

**Comment by viniciusfbb** [comment]
Yes, currently, you should include the dll sk4d.dll with your dll. You don't need to do any manual work, because when you enable skia in your dll project (right click in project and “Enable Skia”), whenever you compile, the skia dll will appear in the output folder along with yours.

**Reply by War3Evo** [reply]
What do I need to do to embed it into the dll?

**Reply by viniciusfbb** [reply]
You can't. It is a shared library. The only way to be able to do this, and the correct way, is to have a library with static linking, so the library's binary object would be included in the application's binary / dll, and it would still reduce the final size because only the features used would be linked and included in the app. However, static linking of C++ libraries in Delphi is said to be impossible on Windows, as the Delphi linker has many limitations with objects generated by other compilers. It's a long story but no one has ever managed to do that. The good news is that we've made a lot of progress on this in the last few months and maybe it's something possible in the near future. :)

**Reply by War3Evo** [reply]
For now, if I include this library with my DLL.... does it need to be in the same directory?  **If not**, how do I tell it which directory the Sdk4Delphi library is in?

**Reply by checkdigits** [reply]
You should include it in the same folder as your app/dll to ensure it's found at runtime. 

It's a shared library which means there might be multiple versions on the same machine. Your app needs to be able to find the version it was compiled against and the best way to do that is put the dll where your app's exe is loaded (a similar process for dlls too).

You can read more about it here: [https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-desktop-applications](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-desktop-applications)

**Reply by War3Evo** [reply]
Let me clarify.  I'm now working on creating VST3 plugins for DAWs.  A VST3 is a renamed dll library.  It's a plug-in for a DAW (Digital Audio Workstation).  It's written in C++, but the project I'm working on has ported it to Delphi.  Ruud Ermers did the majority of the work.  I'm now creating open source plug-ins based on his work and will update the engine code as needed.

After making a few plugins, I'll convert the code into a fully operational drag and drop non-visual registered component for Delphi.  As of right now, my understanding is this code is limited and does not have the full functionality of VST3 SDK.  That's why I am making plugins, to test the limits and expand the code.

https://github.com/War3Evo/RMSVST3/tree/Delphi-11.0

I've not ever worked with DLLs at all until this year, at least as far as I know.

Since VST3 are all found in the same folder in most DAWs, that could be a problem if another person creates a VST3 plug-in with a different version of skia dll.  I could advise them to put them both into another folder.  This could be a really big problem.  If another skia dll version exists in the same VST3 plug-in folder, now what??

Is there a way to force my dll to call skia dll library and NOT look in the same folder??

Maybe even embedding the dll, extract it to another folder, then have it call that dll.  I understand we can't currently statically link, but isn't it possible to embed the file and extract it?  I've got code somewhere that actually does this for executables, and it should work for DLLs too?

**Reply by viniciusfbb** [reply]
Officially there is not much to do. The good news is that we only update the binaries 1x or 2x a year. Also, you can change the Skia.API.pas to load from a subfolder. Storing the dll in a resource and extracting it is not a good idea. The definitive solution for you would be static linking, but it's something you still have to wait a little longer.

**Reply by checkdigits** [reply]
I also want to add a comment here too saying that you should **not** try and embed the DLL as a resource and then extract. It's a very bad idea. Some systems - not all - will detect the extracted DLL as potentially malicious behavior and will automatically quarantine the DLL or continually scan the DLL every time you extract it to ensure it doesn't have a virus or trojan embedded in it, You can, of course, get the extensions folder added to the antivirus exclusion list but this requires the co-operation of the user, antivirus software (and all future updates not to reset it) and so on. Also, extracting the DLL requires your extension to have full read/write access to that folder, which might not always be the case. An installer circumnavigates this restriction by temporarily running with elevated permissions which allows it to install your extension and the Skia dll into the required folder, even if it's normally protected or read-only in normal use by the user.

So, in summary - don't embed the DLL and try to extract it. 😃
